internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -624,12 +616,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 	}
 
-	if r.kindObserver != nil {
-		var gvks []schema.GroupVersionKind
-		for _, ref := range xr.GetResourceReferences() {
-			gvks = append(gvks, ref.GroupVersionKind())
-		}
-		r.kindObserver.WatchComposedResources(gvks...)
+	ws := make([]engine.Watch, len(xr.GetResourceReferences()))
+	for i, ref := range xr.GetResourceReferences() {
+		ws[i] = engine.WatchFor(composed.New(composed.FromReference(ref)), r.watchHandler)
+	}
+	if err := r.engine.StartWatches(r.controllerName, ws...); err != nil {
+		// TODO(negz): If we stop polling this will be a more serious error.
+		log.Debug("Cannot start watches for composed resources. Relying on polling to know when they change.", "controller-name", r.controllerName)
 	}
```

Comment:
I guess I will read it further down, but maybe worth to doc here: do we have to keep calling `StartWatches` for some GC TTL to not apply? Or how does the engine know what to gc?

Reply:
I've added a comment to clarify that the controller engine starts a garbage collector of watches when it starts this controller. I chose not to explain (here) how GC works, mostly because I don't want this comment to get forgotten about it we change how GC works in future. I did add a comment to the GC implementation explaining how it works though.